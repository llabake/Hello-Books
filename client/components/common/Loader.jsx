import React from 'react';

import ajaxLoader from '../../media/ajax-loader.gif';

const Loader = () => {
  return (
    <div
      className="center-align"
      style={{ marginTop: '1em', marginBottom: '1em' }}
    >
      <img src={ajaxLoader} alt="Loading..." />
    </div>
  )
}

export default Loader;