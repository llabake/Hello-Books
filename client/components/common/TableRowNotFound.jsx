import React from 'react';
import PropTypes from 'prop-types';

const TableRowNotFound = ({ text }) => {
  return (
    <tr>
      <td>
      {text}
      </td>
    </tr>
  )
};

TableRowNotFound.propTypes = {
  text: PropTypes.string
};

export default TableRowNotFound;
