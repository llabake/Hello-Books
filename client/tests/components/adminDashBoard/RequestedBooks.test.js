import React from 'react';

import {
  RequestedBooks,
  mapStateToProps,
  mapDispatchToProps
} from '../../../components/adminDashBoard/RequestedBooks';

import adminData from '../../mocks/adminData';
import userData from '../../mocks/userData';
import TableRowNotFound from '../../../components/common/TableRowNotFound';
import { Pagination } from 'react-materialize';
import Loader from '../../../components/common/Loader';


function setup() {
  const props = {
    dispatch: jest.fn(),
    user: adminData.admin,
    errors: {},
    handleSelectedPage: jest.fn(),
    loading: false,
    suggestedBookCount: 0,
    suggestedBooks: [ adminData.suggestion1, adminData.suggestion2 ],
    fetchSuggestedBooks: jest.fn(),
  };

  const state = {
    maxItems: 2,
    showPagination: false,
    activePage: 1,

  };
  const wrapper = shallow(<RequestedBooks {...props} {...state} />);

  return {
    wrapper,
    props,
    state,
  };
}

describe('The Requested books', () => {
  test('should render and not crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  })

  test('should call componentDidMount', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });
    spy = jest.spyOn(RequestedBooks.prototype, 'componentDidMount')
    const { wrapper, props, state } = setup();
    expect(RequestedBooks.prototype.componentDidMount).toHaveBeenCalledTimes(1)
  })

  test('should render the requested books list', () => {
    const { wrapper, props, state } = setup();
    expect(wrapper.find(Pagination).length).toBe(0);
  })

  test('should render nothing found component', () => {
    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      handleSelectedPage: jest.fn(),
      loading: false,
      suggestedBookCount: 0,
      suggestedBooks: [],
      fetchSuggestedBooks: jest.fn(),
    }
    const wrapper = shallow(<RequestedBooks {...props} />);

    expect(wrapper.find(TableRowNotFound).length).toBe(1)

  })

  test('should call handle selected page', () => {
    const { wrapper, props, state } = setup();
    const page = 2
    wrapper.setState({
      activePage: page,
    });
    wrapper.instance().handleSelectedPage(page)
    expect(wrapper.instance().state.activePage).toBe(2);
  });

  test('should call componentWillReceiveProps', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });

    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      handleSelectedPage: jest.fn(),
      loading: true,
      suggestedBookCount: 0,
      suggestedBooks: [],
      fetchSuggestedBooks: jest.fn(),
    }

    const nextProps = {
      dispatch: jest.fn(),
      user: userData.admin,
      errors: {},
      handleSelectedPage: jest.fn(),
      loading: false,
      suggestedBookCount: 1,
      suggestedBooks: [ adminData.suggestion1 ],
      fetchSuggestedBooks: jest.fn(),
    }
    spy = jest.spyOn(RequestedBooks.prototype, 'componentWillReceiveProps');
    const wrapper = shallow(<RequestedBooks {...props} />)
    expect(wrapper.find(Loader).length).toBe(1)
    
    wrapper.setProps(nextProps);
    expect(RequestedBooks.prototype.componentWillReceiveProps).toHaveBeenCalledTimes(1)
    expect(spy).toHaveBeenCalled();
  })

  test('mapDispatchToProps', () => {
    const dispatch = jest.fn();

    expect(mapDispatchToProps(dispatch)).toHaveProperty('fetchSuggestedBooks');

    const { fetchSuggestedBooks } = mapDispatchToProps(dispatch);
    fetchSuggestedBooks();
    expect(dispatch).toHaveBeenCalled();
  });

  test('mapStateToProps', () => {
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('suggestedBooks');
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('loading');
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('suggestedBookCount');
  });

})