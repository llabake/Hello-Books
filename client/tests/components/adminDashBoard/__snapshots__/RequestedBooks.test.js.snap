// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Requested books should render and not crash 1`] = `
<div
  className="col s12 m3  profile-bio"
>
  <div
    className="card-panel"
  >
    <table
      className="bordered centered highlight responsive-table"
    >
      <thead>
        <tr>
          <th
            id="title"
          >
            Requested Books
          </th>
        </tr>
      </thead>
    </table>
    <table
      className="bordered centered highlight responsive-table"
    >
      <tbody>
        <TableRowNotFound
          text="There are no suggested books"
        />
      </tbody>
    </table>
  </div>
</div>
`;
